educe/stac/oneoff/weave.py

```diff
@@ -0,0 +1,267 @@
+text) with those in a 'target' document. This involves copying missing
+annotations over and shifting the text spans of any matching documents
+"""
+
+from collections import namedtuple
+
+from educe.annotation import Span
+from educe.stac.context import enclosed
+
+
+class WeaveException(Exception):
+
+    """
+    Unexpected alignment issues between the source and target
+    document
+    """
+
+    def __init__(self, *args, **kw):
+        super(WeaveException, self).__init__(*args, **kw)
+
+
+class Updates(namedtuple('Updates',
+                         ['shift_if_ge',
+                          'abnormal_src_only',
+                          'abnormal_tgt_only',
+                          'expected_src_only'])):
+
+    """Expected updates to the target document.
+
+    We expect to see four types of annotation:
+
+    1. target annotations for which there exists a
+       source annotation in the equivalent span
+
+    2. target annotations for which there is no equivalent
+       source annotation (eg. Resources, Preferences, but also
+       annotation moves)
+
+    3. source annotations for which there is at least one target
+       annotation at the equivalent span (the mirror to case 1;
+       note that these are not represented in this structure
+       because we don't need to say much about them)
+
+    4. source annotations for which there is no match in the target side
+
+    5. source annotations that lie in between the matching bits of
+       text
+
+    Parameters
+    ----------
+    shift_if_ge : dict(int, int)
+        (case 1 and 2) shift points and offsets for characters
+        in the target document (see `shift_spans`)
+    abnormal_tgt_only : [Annotation]
+        (case 2) annotations that only occur
+        in the target document (weird, found in matches)
+    abnormal_src_only: [Annotation]
+        (case 4) annotations that only occur in the
+        source document (weird, found in matches)
+    abnormal_src_only [Annotation]
+        (case 5) annotations that only occur in the
+        source doc (ok, found in gaps)
+    """
+    def __new__(cls):
+        return super(Updates, cls).__new__(cls, {}, [], [], [])
+
+    def map(self, fun):
+        """
+        Return an `Updates` in which a function has been applied to
+        all annotations in this one (eg. useful for previewing),
+        and to all spans
+        """
+        supercls = super(Updates, Updates)
+        return supercls.__new__(Updates,
+                                self.shift_if_ge,
+                                [fun(x) for x in self.abnormal_src_only],
+                                [fun(x) for x in self.abnormal_tgt_only],
+                                [fun(x) for x in self.expected_src_only])
+
+# ---------------------------------------------------------------------
+#
+# ---------------------------------------------------------------------
+
+
+def src_gaps(matches):
```

Comment:
I'll let you kill this one after one of us has had a chance to debug this code.  Right now, I'm going to leave it in for symmetry.